test/functional/test_framework/key.py

```diff
@@ -444,3 +519,68 @@ def tweak_add(self, tweak):
         ret = ECKey()
         ret.set(tweaked.to_bytes(32, 'big'), self.compressed)
         return ret
+
+def generate_schnorr_nonce():
+    """Generate a random valid Schnorr nonce."""
+    kp = random.randrange(1, SECP256K1_ORDER)
+    assert(kp != 0)
+    R = SECP256K1.affine(SECP256K1.mul([(SECP256K1_G, kp)]))
+    k = kp if jacobi_symbol(R[1], SECP256K1_FIELD_SIZE) == 1 else SECP256K1_ORDER - kp
+    k_key = ECKey()
+    k_key.set(k.to_bytes(32, 'big'), True)
+    return k_key
+
+def aggregate_schnorr_nonces(Rv):
+    """Construct aggregated musig nonce from individually generated nonces."""
+    for idx, R in enumerate(Rv):
+        R_agg = R.p if idx == 0 else SECP256K1.add(R_agg, R.p)
+    R_agg_affine = SECP256K1.affine(R_agg)
+    if jacobi_symbol(R_agg_affine[1], SECP256K1_FIELD_SIZE) != 1:
+        R_agg = SECP256K1.mul([(R_agg, SECP256K1_ORDER - 1)])
+        negated = True
+    else:
+        negated = False
+    ret = ECPubKey()
+    ret.p = R_agg
+    ret.valid = True
+    ret.compressed = True
+    return ret, negated
+
+def generate_musig_key(PubkeyList):
+    """Aggregate individually generated public keys."""
+    pkv = [int.from_bytes(key.get_bytes()[1:], 'big') for key in PubkeyList]
+    pkv.sort()
+    L = b''
+    for px in pkv:
+        L += px.to_bytes(32, 'big')
+    Lh = hashlib.sha256(L).digest()
+    musig_c = {}
+    musig_pk = ECPubKey()
+    for key in PubkeyList:
+        musig_c[key] = hashlib.sha256(Lh + key.get_bytes()[1:]).digest()
+        c = int.from_bytes(musig_c[key],'big')
+        pc = SECP256K1.mul([(key.p, c)])
+        if key.get_bytes() == PubkeyList[0].get_bytes():
+            musig_pk.p = pc
+            musig_pk.valid = True
+            musig_pk.compressed = True
+        else:
+            musig_pk.p = SECP256K1.add(musig_pk.p, pc)
+    return musig_c, musig_pk
+
+def aggregate_musig_signatures(sigs):
+    """Construct valid Schnorr signature from individually generated musig signatures."""
+    assert(sigs)
+    for idx, sig in enumerate(sigs):
+        s = sig[32:]
+        r_x_data = sig[:32]
+        assert(len(s)==32 and len(r_x_data)==32)
```

Comment:
`len(s)` is guaranteed to be 32 by construction. It would be better to assert that `len(sig) == 64` at the top.